File: quickstarts/CS/EventBroker/Src/Stoplight/ServiceImplementations/RealTimeTimer.cs
//===============================================================================
// Microsoft patterns & practices
// Unity Application Block
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
//===============================================================================

using System;
using System.Windows.Forms;
using SimpleEventBroker;
using StopLight.ServiceInterfaces;

namespace StopLight.ServiceImplementations
{
	internal class RealTimeTimer : IStoplightTimer
	{
		private Timer timer;

        [Publishes("TimerTick")]
		public event EventHandler Expired;

		public RealTimeTimer()
		{
			timer = new Timer();
			timer.Tick += OnTick;
		}

		public TimeSpan Duration
		{
			get { return TimeSpan.FromMilliseconds(timer.Interval); }
			set { timer.Interval = (int)value.TotalMilliseconds; }
		}

		public void Start()
		{
			timer.Start();
		}

		private void OnTick(object sender, EventArgs e)
		{
			timer.Stop();
			OnExpired(this);
		}

		protected virtual void OnExpired(object sender)
		{
			EventHandler handlers = Expired;
			if(handlers != null)
			{
				handlers(this, EventArgs.Empty);
			}
		}
	}
}

File: quickstarts/VB/EventBroker/Src/Stoplight/ServiceImplementations/RealTimeTimer.vb
'===============================================================================
' Microsoft patterns & practices
' Unity Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports System
Imports System.Windows.Forms
Imports SimpleEventBroker
Imports StopLight.ServiceInterfaces

Namespace StopLight.ServiceImplementations
	Friend Class RealTimeTimer
		Implements IStoplightTimer
    Private _timer As Timer

    <Publishes("TimerTick")> _
    Public Event Expired As EventHandler Implements IStoplightTimer.Expired

    Public Sub New()
      _timer = New Timer()
      AddHandler _timer.Tick, AddressOf OnTick
    End Sub

    Public Property Duration() As TimeSpan Implements IStoplightTimer.Duration
      Get
        Return TimeSpan.FromMilliseconds(_timer.Interval)
      End Get
      Set(ByVal value As TimeSpan)
        _timer.Interval = CType(value.TotalMilliseconds, Integer)
      End Set
    End Property

    Public Sub Start() Implements IStoplightTimer.Start
      _timer.Start()
    End Sub

    Private Sub OnTick(ByVal sender As Object, ByVal e As EventArgs)
      _timer.[Stop]()
      OnExpired(Me)
    End Sub

		Protected Overridable Sub OnExpired(ByVal sender As Object)
      Dim handlers As EventHandler = ExpiredEvent
      If Not handlers Is Nothing Then
        RaiseEvent Expired(Me, EventArgs.Empty)
      End If
		End Sub
	End Class
End Namespace

